src/definitions/wrapping.ts

```diff
@@ -164,127 +172,158 @@ export function isNexusPlugin(obj: any): obj is NexusPlugin {
-export function wrapAsNexusType(
-  baseType: AllNexusNamedTypeDefs | string,
-  wrapping: NexusWrapKind[]
-): AllNexusTypeDefs {
-  return wrapping.reduce<any>((type, kind) => {
-    if (kind.type === 'NamedType' && kind.nullable === false) {
-      return nonNull(type)
+/**
+ * Apply the wrapping consistently to the arg `type`
+ *
+ * nonNull(list(stringArg())) -> arg({ type: nonNull(list('String')) })
+ */
+export function normalizeArgWrapping(argVal: AllNexusArgsDefs): NexusArgDef<AllInputTypes> {
+  if (isNexusArgDef(argVal)) {
+    return argVal
+  }
+  if (isNexusWrappingType(argVal)) {
+    let { namedType, wrapping } = unwrapNexusDef(argVal)
+    if (isNexusArgDef(namedType)) {
+      const config = namedType.value
+      return arg({ ...config, type: applyNexusWrapping(config.type, wrapping) })
     }
+    return arg({ type: applyNexusWrapping(namedType, wrapping) })
+  }
+  return arg({ type: argVal })
+}
 
-    if (kind.type === 'List' && kind.nullable === true) {
-      return list(type)
+/**
+ * Applies the ['List', 'NonNull', 'Nullable']
+ * @param toWrap
+ * @param wrapping
+ */
+export function applyNexusWrapping(toWrap: any, wrapping: NexusWrapKind[]) {
+  let finalType = toWrap
+  wrapping.forEach((wrap) => {
+    if (wrap === 'List') {
+      finalType = list(finalType)
+    } else if (wrap === 'NonNull') {
+      finalType = nonNull(finalType)
+    } else if (wrap === 'Null') {
+      finalType = nullable(finalType)
+    } else {
+      throw new Unreachable(wrap)
     }
+  })
+  return finalType
+}
 
-    if (kind.type === 'List' && kind.nullable === false) {
-      return nonNull(list(type))
+/**
+ * Takes the "nonNullDefault" value, the chained wrapping, and the field wrapping,
+ * to determine the proper list of wrapping to apply to the field
+ */
+export function finalizeWrapping(
+  location: string,
+  nonNullDefault: boolean,
+  typeWrapping: NexusWrapKind[] | ReadonlyArray<NexusWrapKind>,
+  chainWrapping?: NexusWrapKind[]
+): NexusFinalWrapKind[] {
+  let finalChain: NexusFinalWrapKind[] = []
+  if (typeWrapping.length && chainWrapping?.length) {
+    throw new Error(
+      `Cannot use t.list / nonNull chaining and list() / nonNull() type wrapping the same time (on ${location})`
+    )
+  }
+  const allWrapping = typeWrapping.concat(chainWrapping ?? [])
+  // Ensure the first item is wrapped, if we're not guarding
+  if (nonNullDefault && (!allWrapping[0] || allWrapping[0] === 'List')) {
+    allWrapping.unshift('NonNull')
+  }
+  for (let i = 0; i < allWrapping.length; i++) {
+    const current = allWrapping[i]
+    const next = allWrapping[i + 1]
+    if (current === 'Null') {
+      continue
+    } else if (current === 'NonNull') {
+      finalChain.push('NonNull')
+    } else if (current === 'List') {
+      finalChain.push('List')
+      if (nonNullDefault && (next === 'List' || !next)) {
+        finalChain.push('NonNull')
+      }
     }
-
-    return type
-  }, baseType)
+  }
+  return finalChain
 }
```

Comment:
Errrr. I swear I tried the very same approach cause the other approach was terribly ugly but didn't manage to have it working 😫. What I wasn't doing properly when I tried was to keep reading the wrapping by pairs when inserting back the `NonNull`'s 🤦‍♂️

Reply:
Haha. To be fair this took me like two hours of staring at it and trying to figure out what it should be doing. Wouldn’t have been able to get it without the existing test suites you added.